File: DigitalImageProcessing/cpp/hough_transform.cpp
/**
 * @File    : hough_transform.cpp
 * @Brief   : 霍夫变换 (Hough Transform) 进行二值图像的直线检测
 * @Author  : Wei Li
 * @Date    : 2021-10-18
*/


#include <iostream>
#include <vector>
#include <map>
#include <utility>
#include <cmath>
#include <opencv2/core.hpp>
#include <opencv2/imgproc.hpp>
#include <opencv2/highgui.hpp>


std::map<std::vector<int>, std::vector<cv::Point>> HoughTransformLine(cv::Mat image, cv::Mat &accumulator, float stepTheta=1, float stepRho=1)
{
    int rows = image.rows;
    int cols = image.cols;
    // 可能出现的最大垂线的长度
    int L = std::round(std::sqrt(std::pow(rows - 1, 2.0) + std::pow(cols - 1, 2.0))) + 1;
    // 初始化投票器
    int numTheta = int(180.0 / stepTheta);
    int numRho = int(2 * L / stepRho + 1);
    accumulator = cv::Mat::zeros(cv::Size(numTheta, numRho), CV_32SC1);

    // 初始化类 map ，存储共线的点
    std::map<std::vector<int>, std::vector<cv::Point>> lines;
    for (size_t i = 0; i < numRho; ++i)
    {
        for (size_t j = 0; i < numTheta; ++j)
        {
            lines.insert(std::make_pair(std::vector<int>(j, i), std::vector<cv::Point>()));
        }
    }
    // 投票计数
    for (size_t y = 0; y < rows; ++y)
    {
        for (size_t x = 0; x < cols; ++x)
        {
            if (image.at<uchar>(cv::Point(x, y)) == 255)
            {
                for (size_t m = 0; m < numTheta; ++m)
                {
                    // 对每一个角度，计算对应的 rho 值 
                    float rho1 = x * std::cos(stepTheta * m / 180.0 * CV_PI);
                    float rho2 = y * std::sin(stepTheta * m / 180.0 * CV_PI);
                    float rho = rho1 + rho2;
                    // 计算投票到哪一个区域
                    int n = int(std::round(rho + L) / stepRho);
                    // 累加 1
                    accumulator.at<int>(n, m) += 1;
                    // 记录该点
                    lines.at(std::vector<int>(m, n)).push_back(cv::Point(x, y));
                }
            }
        }
    }
    return lines;
}


// -------------------------------
int main(int argc, char **argv)
{
    if (argc > 1)
    {
        cv::Mat image = cv::imread(argv[1], 0);
        if (!image.data)
        {
            std::cout << "Error: reading image unsuccesfully." << std::endl;
            return -1;
        }
        cv::imshow("OriginImage", image);

        // 边缘二值图
        cv::Mat edge;
        cv::Canny(image, edge, 50, 200);

        // ------------ Hough Transform Line ------------
        cv::Mat accu;
        std::map<std::vector<int>, std::vector<cv::Point>> lines;
        lines = HoughTransformLine(edge, accu);

        // 投票器的灰度级可视化
        double maxValue;
        cv::minMaxLoc(accu, nullptr, &maxValue, nullptr, nullptr);
        cv::Mat grayAccu;
        accu.convertTo(grayAccu, CV_32FC1, 1.0 / maxValue);
        cv::imshow("AccumulatorGray", grayAccu);
        // 绘制投票数量大于某一阈值的直线
        int vote = 150;
        for (size_t r = 1; r < accu.rows - 1; ++r)
        {
            for (size_t c = 1; c < accu.cols - 1; ++c)
            {
                int current = accu.at<int>(r, c);
                if (current > vote)
                {
                    int left_top = accu.at<int>(r - 1, c - 1);
                    int top = accu.at<int>(r - 1, c);
                    int right_top = accu.at<int>(r - 1, c + 1);
                    int left = accu.at<int>(r, c - 1);
                    int right = accu.at<int>(r, c + 1);
                    int left_bottom = accu.at<int>(r + 1, c - 1);
                    int bottom = accu.at<int>(r - 1, c);
                    int right_bottom = accu.at<int>(r + 1, c + 1);
                    // 判断该位置是不是局部最大值
                    if (current > left && current > top && current > right_top && current > left && current < right && current > left_bottom && current > bottom && current > right_bottom)
                    {
                        std::vector<cv::Point> line = lines.at(std::vector<int>(c, r));
                        int s = line.size();
                        cv::line(image, line.at(0), line.at(s - 1), cv::Scalar(255), 2);
                    }
                }
            }
        }

        cv::waitKey(0);
        cv::destroyAllWindows();
        return 0;
    }
    else
    {
        std::cout << "Usage: OpenCV gauss convolution image." << std::endl;
        return -1;
    }
}

File: DigitalImageProcessing/python/hough_transform.py
#!/usr/bin/env python3
# encoding: utf-8

"""
@Funciton: 霍夫变换 (Hough Transform) 进行二值图像的直线检测
@Python Version: 3.8
@Author: Wei Li
@Date: 2021-10-18
"""

import sys
import math
import numpy as np
import matplotlib.pyplot as plt
from mpl_toolkits.mplot3d import Axes3D
import cv2 as cv


def HoughTransformLine(image, stepTheta=1, stepRho=1):
    """实现标准的霍夫直线检测

    Args:
        image ([ndarray]): 输入待检测的二值图像
        stepTheta (int, optional): 计算机程序实现原理时候的离散化参数 θ. Defaults to 1.
        stepRho (int, optional): 计算机程序实现原理时候的离散化参数 ρ. Defaults to 1.

    Returns:
        [tuple]: (计数器，对应的哪些点是共线的)
    """
    rows, cols = image.shape
    # 图像中可能出现的最大垂线的长度
    L = round(math.sqrt(pow(rows - 1, 2.0) + pow(cols - 1, 2.0))) + 1
    # 初始化投票器
    numTheta = int(180.0 / stepTheta)
    numRho = int(2 * L / stepRho + 1)
    accumulator = np.zeros((numRho, numTheta), np.int32)
    # 建立字典, 对应的那些点是共线的
    accuDict = {}
    for k1 in range(numRho):
        for k2 in range(numTheta):
            accuDict[(k1, k2)] = []

    # 投票计数
    for y in range(rows):
        for x in range(cols):
            if (image[y][x] == 255):  # 只对边缘点做霍夫变换
                for m in range(numTheta):
                    # 对每一个角度，计算对应的 rho 值
                    rho = x * math.cos(stepTheta * m / 180.0 * math.pi) + y * math.sin(stepTheta * m / 180.0 * math.pi)
                    # 计算投票哪一个区域
                    n = int(round(rho + L) / stepRho)
                    # 投票计数 +1
                    accumulator[n, m] += 1
                    # 记录该点
                    accuDict[(n, m)].append((y, x))

    return accumulator, accuDict 

# --------------------------
if __name__ == "__main__":
    if len(sys.argv) > 1:
        image = cv.imread(sys.argv[1], 0)
        if image is None:
            print(f"Error: no such file or dictory.")
        cv.imshow("OriginImage", image)

        # 边缘二值图
        edge_canny = cv.Canny(image, 50, 200)
        cv.imshow("EdgeCanny", edge_canny)

        # -------- Hough Transform Line --------
        # -------- 对计数器进行三维展示和二值化展示 --------
        accumulator, accuDict = HoughTransformLine(edge_canny, 1, 1)

        # 计算器的二维直方图展示
        rows, cols = accumulator.shape
        fig = plt.figure()
        plt.gca(projection="3d")
        ax = fig.add_subplot(projection='3d')
        X, Y = np.mgrid[0:rows:1, 0:cols:1]
        surface = ax.plot_wireframe(X, Y, accumulator, cstride=1, rstride=1, color="gray")
        ax.set_xlabel(u"$\\rho$")
        ax.set_ylabel(u"$\\theta$")
        ax.set_zlabel("accumulator")
        ax.set_zlim3d(0, np.max(accumulator))
        plt.show()
        plt.close()

        # 计数器的灰度级显示
        grayAccu = accumulator / float(np.max(accumulator))
        grayAccu = 255 * grayAccu
        grayAccu = grayAccu.astype(np.uint8)
        # 只需要绘制投票数大于 60 的直线
        voteThresh = 60
        for r in range(rows):
            for c in range(cols):
                if accumulator[r][c] > voteThresh:
                    points = accuDict[(r, c)]
                    cv.line(image, points[0], points[len(points) - 1], (255), 2)
        cv.imshow("Accumulator", grayAccu)

        cv.waitKey()
        cv.destroyAllWindows()

    else:
        print("Usage: python hough_transform.py imageFile")
